packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -0,0 +1,295 @@
+ * ```
+ */
+export function FunctionCall({
+  name,
+  partial,
+  args,
+}: {
+  name: string;
+  partial?: boolean;
+  args: Record<string, string | number | boolean | null>;
+}) {
+  return `Call function ${name} with ${partial ? '(incomplete) ' : ''}${JSON.stringify(args)}`;
+}
+
+/**
+ * Renders to the output of a previous {@link FunctionCall} component, for use within a {@link ChatCompletion}.
+ *
+ * See https://platform.openai.com/docs/guides/gpt/chat-completions-api for more detail.
+ *
+ * @example
+ * ```tsx
+ *    <ChatCompletion>
+ *      <UserMessage>What is 258 * 322?</UserMessage>
+ *      <FunctionCall name="evaluate_math" args={expression: "258 * 322"} />
+ *      <FunctionResponse name="evaluate_math">83076</FunctionResponse>
+ *    </ChatCompletion>
+ *
+ *    ==> "That would be 83,076."
+ * ```
+ */
+export async function FunctionResponse(
+  { name, children }: { name: string; children: Node },
+  { render }: AI.ComponentContext
+) {
+  const output = await render(children);
+  return `function ${name} returns ${output}`;
+}
+
+interface ConversationMessageType<T, C extends AI.Component<any>> {
+  type: T;
+  element: AI.Element<AI.PropsOfComponent<C>>;
+}
+
+/**
+ * A type that represents a conversation message.
+ */
+export type ConversationMessage =
+  | ConversationMessageType<'user', typeof UserMessage>
+  | ConversationMessageType<'assistant', typeof AssistantMessage>
+  | ConversationMessageType<'system', typeof SystemMessage>
+  | ConversationMessageType<'functionCall', typeof FunctionCall>
+  | ConversationMessageType<'functionResponse', typeof FunctionResponse>;
+
+/** @hidden */
+export function isConversationalComponent(element: AI.Element<any>): boolean {
+  return (
+    element.tag === UserMessage ||
+    element.tag === AssistantMessage ||
+    element.tag === SystemMessage ||
+    element.tag === FunctionCall ||
+    element.tag === FunctionResponse
+  );
+}
+
+function toConversationMessages(partialRendering: AI.PartiallyRendered[]): ConversationMessage[] {
+  return partialRendering
+    .flatMap((e) => (AI.isElement(e) ? [e] : []))
+    .map<ConversationMessage>((e) => {
+      switch (e.tag) {
+        case UserMessage:
+          return { type: 'user', element: e };
+        case AssistantMessage:
+          return { type: 'assistant', element: e };
+        case SystemMessage:
+          return { type: 'system', element: e };
+        case FunctionCall:
+          return { type: 'functionCall', element: e };
+        case FunctionResponse:
+          return { type: 'functionResponse', element: e };
+        default:
+          throw new AIJSXError(
+            `Unexpected tag (${e.tag.name}) in conversation`,
+            ErrorCode.UnexpectedPartialRenderResult,
+            'internal'
+          );
+      }
+    });
+}
+
+/** @hidden */
+export async function renderToConversation(
+  conversation: AI.Node,
+  render: AI.ComponentContext['render']
+): Promise<ConversationMessage[]> {
```

Comment:
```suggestion
) {
```

Is this type annotation necessary?